File: windows/ClaudeUsage/PixelFont.cs
using System.Drawing;

namespace ClaudeUsage;

/// <summary>
/// 3x5 pixel digits. Percentages are drawn with these rather than a system font; the panel's
/// pixel-art look only holds together when the numbers are built from blocks too.
/// </summary>
internal static class PixelFont
{
    private const int GlyphWidth = 3;
    private const int GlyphHeight = 5;
    private const int Spacing = 1;   // letter spacing, in cells

    private static readonly Dictionary<char, string[]> Glyphs = new()
    {
        ['0'] = ["XXX", "X.X", "X.X", "X.X", "XXX"],
        ['1'] = ["..X", "..X", "..X", "..X", "..X"],
        ['2'] = ["XXX", "..X", "XXX", "X..", "XXX"],
        ['3'] = ["XXX", "..X", "XXX", "..X", "XXX"],
        ['4'] = ["X.X", "X.X", "XXX", "..X", "..X"],
        ['5'] = ["XXX", "X..", "XXX", "..X", "XXX"],
        ['6'] = ["XXX", "X..", "XXX", "X.X", "XXX"],
        ['7'] = ["XXX", "..X", "..X", "..X", "..X"],
        ['8'] = ["XXX", "X.X", "XXX", "X.X", "XXX"],
        ['9'] = ["XXX", "X.X", "XXX", "..X", "XXX"],
        ['%'] = ["X.X", "..X", ".X.", "X..", "X.X"],
        ['.'] = ["...", "...", "...", "...", "..X"],
        [','] = ["...", "...", "...", "..X", ".X."],
        ['$'] = [".X.", "XXX", "XX.", ".XX", "XXX"],
        [' '] = ["...", "...", "...", "...", "..."],
    };

    public static int Height(int cell) => GlyphHeight * cell;

    public static int Width(string text, int cell) =>
        text.Length == 0 ? 0 : text.Length * GlyphWidth * cell + (text.Length - 1) * Spacing * cell;

    public static void Draw(Graphics graphics, string text, int x, int y, int cell, Color color)
    {
        using var brush = new SolidBrush(color);
        int cursor = x;

        foreach (char character in text)
        {
            if (!Glyphs.TryGetValue(character, out string[]? glyph))
            {
                cursor += (GlyphWidth + Spacing) * cell;
                continue;
            }

            for (int row = 0; row < GlyphHeight; row++)
                for (int column = 0; column < GlyphWidth; column++)
                    if (glyph[row][column] == 'X')
                        graphics.FillRectangle(brush, cursor + column * cell, y + row * cell, cell, cell);

            cursor += (GlyphWidth + Spacing) * cell;
        }
    }
}

File: macos/Sources/ClaudeLimitBar/PixelFont.swift
import AppKit

// 3x5 pixel digits, the same glyph table the Windows app draws from.
//
// Percentages are built from blocks rather than set in a system font for the
// same reason the bars are square: a pixel-art panel stops reading as one the
// moment its largest element is a smoothly antialiased typeface. Only the
// headline figures use this — labels and reset times stay in the system font,
// exactly as on Windows, because five-pixel-tall prose is unreadable.
enum PixelFont {
    private static let glyphWidth = 3
    private static let glyphHeight = 5
    private static let spacing = 1  // letter spacing, in cells

    private static let glyphs: [Character: [String]] = [
        "0": ["XXX", "X.X", "X.X", "X.X", "XXX"],
        "1": ["..X", "..X", "..X", "..X", "..X"],
        "2": ["XXX", "..X", "XXX", "X..", "XXX"],
        "3": ["XXX", "..X", "XXX", "..X", "XXX"],
        "4": ["X.X", "X.X", "XXX", "..X", "..X"],
        "5": ["XXX", "X..", "XXX", "..X", "XXX"],
        "6": ["XXX", "X..", "XXX", "X.X", "XXX"],
        "7": ["XXX", "..X", "..X", "..X", "..X"],
        "8": ["XXX", "X.X", "XXX", "X.X", "XXX"],
        "9": ["XXX", "X.X", "XXX", "..X", "XXX"],
        "%": ["X.X", "..X", ".X.", "X..", "X.X"],
        ".": ["...", "...", "...", "...", "..X"],
        ",": ["...", "...", "...", "..X", ".X."],
        "$": [".X.", "XXX", "XX.", ".XX", "XXX"],
        "+": ["...", ".X.", "XXX", ".X.", "..."],
        "-": ["...", "...", "XXX", "...", "..."],
        " ": ["...", "...", "...", "...", "..."],
    ]

    static func height(cell: CGFloat) -> CGFloat { CGFloat(glyphHeight) * cell }

    static func width(_ text: String, cell: CGFloat) -> CGFloat {
        guard !text.isEmpty else { return 0 }
        let count = CGFloat(text.count)
        return count * CGFloat(glyphWidth) * cell + (count - 1) * CGFloat(spacing) * cell
    }

    // Draws top-down from (x, y), which is why the view this lives in is
    // flipped. An unknown character advances the cursor without drawing, so a
    // string the table doesn't fully cover still lines up.
    static func draw(_ text: String, at point: NSPoint, cell: CGFloat, color: NSColor) {
        color.setFill()
        var cursor = point.x

        for character in text {
            guard let glyph = glyphs[character] else {
                cursor += CGFloat(glyphWidth + spacing) * cell
                continue
            }

            for (rowIndex, row) in glyph.enumerated() {
                for (columnIndex, pixel) in row.enumerated() where pixel == "X" {
                    NSRect(
                        x: cursor + CGFloat(columnIndex) * cell,
                        y: point.y + CGFloat(rowIndex) * cell,
                        width: cell, height: cell
                    ).fill()
                }
            }

            cursor += CGFloat(glyphWidth + spacing) * cell
        }
    }
}
